tools/state-parts-dump-check/src/cli.rs

```diff
@@ -0,0 +1,471 @@
+            let rpc_client = crate::rpc_requests::RpcClient::new(&chain_id);
+            let handle = thread::spawn(move || 
+                run_loop_single_shard(
+                    chain_id,
+                    root_dir,
+                    s3_bucket,
+                    s3_region,
+                    gcs_bucket,
+                    shard_id,
+                    rpc_client
+                )               
+            );
+            handles.push(handle);
+        }
+
+        for handle in handles {
+            let _ = handle.join().unwrap();
+        }
+        
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct DumpCheckIterInfo {
+    pub latest_block_hash: String,
+    pub latest_epoch_id: String,
+    pub prev_epoch_id: String,
+    pub prev_epoch_height: u64,
+}
+
+pub fn create_external_connection(
+    root_dir: Option<PathBuf>,
+    bucket: Option<String>,
+    region: Option<String>,
+    gcs_bucket: Option<String>,
+) -> ExternalConnection {
+    if let Some(root_dir) = root_dir {
+        ExternalConnection::Filesystem { root_dir }
+    } else if let (Some(bucket), Some(region)) = (bucket, region) {
+        let bucket = create_bucket_readonly(&bucket, &region, Duration::from_secs(5))
+            .expect("Failed to create an S3 bucket");
+        ExternalConnection::S3 { bucket: Arc::new(bucket) }
+    } else if let Some(bucket) = gcs_bucket {
+        ExternalConnection::GCS {
+            gcs_client: Arc::new(cloud_storage::Client::default()),
+            reqwest_client: Arc::new(reqwest::Client::default()),
+            bucket,
+        }
+    } else {
+        panic!(
+            "Please provide --root-dir, or both of --s3-bucket and --s3-region, or --gcs-bucket"
+        );
+    }
+}
+
+fn validate_state_part(state_root: &StateRoot, part_id: PartId, part: &[u8]) -> bool {
+    match BorshDeserialize::try_from_slice(part) {
+        Ok(trie_nodes) => {
+            match Trie::validate_state_part(state_root, part_id, trie_nodes) {
+                Ok(_) => true,
+                // Storage error should not happen
+                Err(err) => {
+                    tracing::error!(target: "state-parts", ?err, "State part storage error");
+                    false
+                }
+            }
+        }
+        // Deserialization error means we've got the data from malicious peer
+        Err(err) => {
+            tracing::error!(target: "state-parts", ?err, "State part deserialization error");
+            false
+        }
+    }
+}
+
+fn run_loop_single_shard(
+    chain_id: String,
+    root_dir: Option<PathBuf>,
+    s3_bucket: Option<String>,
+    s3_region: Option<String>,
+    gcs_bucket: Option<String>,
+    shard_id: ShardId,
+    rpc_client: crate::rpc_requests::RpcClient,
+) -> anyhow::Result<()> {
+    let mut last_check_status =
+        Ok(StatePartsDumpCheckStatus::WaitingForParts { epoch_height: 0 });
+
+    let mut last_check_iter_info;
+
+    loop {
```

Comment:
Stopping and starting actix repeatedly may be problematic, but I don't know why.
I suggest to start and stop actix only once, here, and not on line 282.

```suggestion
 sys.block_on(async move {
    loop {
```